parsing/parchmint-parser.js

```diff
@@ -208,12 +275,24 @@ class ParchmintParser {
      * @param {object}  compObj     An object with the fields name, id,
      *                              x-span, y-span, and entity.
      * @param {Array}   ports       The list of ports this Component
-     *                              references in the form of a map.
+     *                              references in the form of a map. This
+     *                              parameter can be left empty. If this
+     *                              parameter evaluates to falsy it will not be
+     *                              added to the Component.
+     *
      * @returns {Component} The resulting Component object.
      */
-    static getParsedComponent(compObj, ports) {
-        return new Component(compObj['name'], compObj['id'], compObj['x-span'], compObj['y-span'],
-                compObj['entity'], ports);
+    static getParsedComponent(compObj, ports = null) {
+        let ret = new Component(compObj['name'], compObj['id'], compObj['x-span'], compObj['y-span'],
+                compObj['entity']);
+
+        // There might not be any ports on the layer, so only add it if we have one, otherwise leave it
+        // as the default value
+        if (ports) {
+            ret.ports = ports;
+        }
```

Comment:
We could set it in the constructor using the code you suggested for the above comment, but it couldn't be used for every parameter because our default string, the empty string, equates to falsy in JS.

Reply:
I don't quite understand how the default strings means that we won't be able to set all the values in the parameters (if they exist) and then hide this code in the constructor, can you give me an example or a bit more explanation?

Reply:
I misread [the documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_Operators#Logical_OR). I was under the impression that because the empty string equates to falsy, the logical or statement would return a value of false.

I'll switch all default parameters to the ```value || default_value``` notation.

Reply:
I just read [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters#Examples) that explicitly passing ```undefined``` to a parameter will use its default value. This would allow us to embed the ```Map.get``` calls directly into the constructors because they return ```undefined``` if no value exists.

Reply:
I see the refactor of the code to utilize the constructor rather than building the elements externally, but I don't see any modifications to the constructors themselves. Is that forthcoming or do we need to modify that as well?

Reply:
We shouldn't need to modify the constructors. The default parameters are just shorthand for the ```val || default_val```, so whenever we want the default_val we can just pass in ```undefined```.